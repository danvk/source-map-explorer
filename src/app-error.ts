import { formatPercent } from './helpers';
import { ErrorCode } from './index';

// If we need advanced error consider using https://github.com/joyent/node-verror
export class AppError extends Error {
  code?: ErrorCode;
  cause?: Error;

  constructor(errorContext: ErrorContext, error?: NodeJS.ErrnoException) {
    super();
    // https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work
    Object.setPrototypeOf(this, AppError.prototype);

    const message = getErrorMessage(errorContext);

    this.message = error ? `${message}: ${error.message}` : message;
    this.code = errorContext.code;

    Error.captureStackTrace(this, AppError);
  }
}

export const SOURCE_MAP_INFO_URL =
  'https://github.com/danvk/source-map-explorer/blob/master/README.md#generating-source-maps';

interface CommonErrorContext {
  code: 'NoBundles' | 'NoSourceMap' | 'CannotSaveFile' | 'CannotCreateTempFile' | 'Unknown';
}

interface OneSourceSourceMapErrorContext {
  code: 'OneSourceSourceMap';
  filename: string;
}

interface UnmappedBytesErrorContext {
  code: 'UnmappedBytes';
  totalBytes: number;
  unmappedBytes: number;
}

interface InvalidMappingLineErrorContext {
  code: 'InvalidMappingLine';
  generatedLine: number;
  maxLine: number;
}

interface InvalidMappingColumnErrorContext {
  code: 'InvalidMappingColumn';
  generatedLine: number;
  generatedColumn: number;
  maxColumn: number;
}

interface CannotOpenTempFileErrorContext {
  code: 'CannotOpenTempFile';
  error: Buffer;
  tempFile: string;
}

export type ErrorContext =
  | CommonErrorContext
  | OneSourceSourceMapErrorContext
  | UnmappedBytesErrorContext
  | InvalidMappingLineErrorContext
  | InvalidMappingColumnErrorContext
  | CannotOpenTempFileErrorContext;

export function getErrorMessage(context: ErrorContext): string {
  switch (context.code) {
    case 'NoBundles':
      return 'No file(s) provided';

    case 'NoSourceMap':
      return `Unable to find a source map.
See ${SOURCE_MAP_INFO_URL}`;

    case 'OneSourceSourceMap': {
      return `Your source map only contains one source (${context.filename}).
This can happen if you use browserify+uglifyjs, for example, and don't set the --in-source-map flag to uglify.
See ${SOURCE_MAP_INFO_URL}`;
    }

    case 'UnmappedBytes': {
      const { unmappedBytes, totalBytes } = context;

      const bytesString = formatPercent(unmappedBytes, totalBytes, 2);

      return `Unable to map ${unmappedBytes}/${totalBytes} bytes (${bytesString}%)`;
    }

    case 'InvalidMappingLine': {
      const { generatedLine, maxLine } = context;

      return `Your source map refers to generated line ${generatedLine}, but the source only contains ${maxLine} line(s).
Check that you are using the correct source map.`;
    }

    case 'InvalidMappingColumn': {
      const { generatedLine, generatedColumn, maxColumn } = context;

      return `Your source map refers to generated column ${generatedColumn} on line ${generatedLine}, but the source only contains ${maxColumn} column(s) on that line.
Check that you are using the correct source map.`;
    }

    case 'CannotSaveFile':
      return 'Unable to save HTML to file';

    case 'CannotCreateTempFile':
      return 'Unable to create a temporary HTML file';

    case 'CannotOpenTempFile': {
      const { error, tempFile } = context;

      return `Unable to open web browser. ${error.toString().trim()}
Either run with --html, --json, --tsv, --file, or view HTML for the visualization at:
${tempFile}`;
    }

    default:
      return 'Unknown error';
  }
}
