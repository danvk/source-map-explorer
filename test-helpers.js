const { existsSync } = require('fs');
const spawn = require('cross-spawn');
const concat = require('concat-stream');
const PATH = process.env.PATH;

/*
  `createProcess` and `execute` are from
  https://gist.github.com/zorrodg/c349cf54a3f6d0a9ba62e0f4066f31cb
  https://medium.com/@zorrodg/integration-tests-on-node-js-cli-part-1-why-and-how-fa5b1ba552fe
*/

/**
 * Creates a child process with script path
 * @param {string} processPath Path of the process to execute
 * @param {Array} args Arguments to the command
 */
function createProcess(processPath, args = []) {
  // Ensure that path exists
  if (!processPath || !existsSync(processPath)) {
    throw new Error('Invalid process path');
  }

  args = [processPath].concat(args);

  return spawn('node', args, {
    env: {
      NODE_ENV: 'test',
      preventAutoStart: false,
      PATH // This is needed in order to get all the binaries in your current terminal
    },
    stdio: [null, null, null, 'ipc'] // This enables interprocess communication (IPC)
  });
}

/**
 * Creates a command and executes inputs (user responses) to the stdin.
 * Returns a promise that resolves when all inputs are sent.
 * Rejects the promise if any error.
 * @param {string} processPath Path of the process to execute
 * @param {Array} args Arguments to the command
 */
function execute(processPath, args = []) {
  const childProcess = createProcess(processPath, args);
  childProcess.stdin.setEncoding('utf-8');
  const promise = new Promise((resolve, reject) => {
    childProcess.stderr.once('data', err => {
      reject(err.toString());
    });
    childProcess.on('error', reject);
    childProcess.stdout.pipe(
      concat(result => {
        resolve(result.toString());
      })
    );
  });
  return promise;
}

module.exports = {
  execute
};
