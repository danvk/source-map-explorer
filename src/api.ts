import glob from 'glob';
import { partition, flatMap, isString } from 'lodash';

import { exploreBundle, UNMAPPED_KEY, SOURCE_MAP_COMMENT_KEY } from './explore';
import { AppError, getErrorMessage } from './app-error';
import {
  BundlesAndFileTokens,
  ExploreOptions,
  ExploreResult,
  Bundle,
  ExploreErrorResult,
  ExploreBundleResult,
} from './index';
import { formatOutput, saveOutputToFile } from './output';

/**
 * Analyze bundle(s)
 */
export async function explore(
  bundlesAndFileTokens: BundlesAndFileTokens,
  options: ExploreOptions = {}
): Promise<ExploreResult> {
  bundlesAndFileTokens = Array.isArray(bundlesAndFileTokens)
    ? bundlesAndFileTokens
    : [bundlesAndFileTokens];

  if (bundlesAndFileTokens.length === 0) {
    throw new AppError({ code: 'NoBundles' });
  }

  // Separate bundles from file tokens
  const [fileTokens, bundles] = partition(bundlesAndFileTokens, isString);

  // Get bundles from file tokens
  bundles.push(...getBundles(fileTokens));

  const results = await Promise.all(
    bundles.map(bundle =>
      exploreBundle(bundle, options).catch<ExploreErrorResult>(error =>
        onExploreError(bundle, error)
      )
    )
  );

  const exploreResult = getExploreResult(results, options);

  // Reject if none of results is successful
  if (exploreResult.bundles.length === 0) {
    return Promise.reject(exploreResult);
  }

  saveOutputToFile(exploreResult, options);

  return exploreResult;
}

/**
 * Expand list of file tokens into a list of bundles
 */
export function getBundles(fileTokens: string[]): Bundle[] {
  const filenames = flatMap(fileTokens, filePath =>
    glob.hasMagic(filePath) ? expandGlob(filePath) : filePath
  );

  const [mapFilenames, codeFilenames] = partition(filenames, filename => filename.endsWith('.map'));

  return codeFilenames.map<Bundle>(code => ({
    code,
    map: mapFilenames.find(filename => filename === `${code}.map`),
  }));
}

function expandGlob(pattern: string): string[] {
  // Make sure pattern match `.map` files as well
  if (pattern.endsWith('.js')) {
    pattern = `${pattern}?(.map)`;
  }

  return glob.sync(pattern);
}

export function getBundleName(bundle: Bundle): string {
  return Buffer.isBuffer(bundle.code) ? 'Buffer' : bundle.code;
}

/**
 * Handle error during bundle processing
 */
function onExploreError(bundle: Bundle, error: NodeJS.ErrnoException): ExploreErrorResult {
  return {
    bundleName: getBundleName(bundle),
    code: error.code || 'Unknown',
    message: error.message,
    error,
  };
}

function getExploreResult(
  results: (ExploreBundleResult | ExploreErrorResult)[],
  options: ExploreOptions
): ExploreResult {
  const [bundles, errors] = partition(
    results,
    (result): result is ExploreBundleResult => 'files' in result
  );

  errors.push(...getPostExploreErrors(bundles));

  return {
    bundles,
    errors,
    ...(bundles.length > 0 && { output: formatOutput(bundles, options) }),
  };
}

const SPECIAL_FILENAMES = [UNMAPPED_KEY, SOURCE_MAP_COMMENT_KEY];

function getPostExploreErrors(exploreBundleResults: ExploreBundleResult[]): ExploreErrorResult[] {
  const errors: ExploreErrorResult[] = [];

  const isSingleBundle = exploreBundleResults.length === 1;

  for (const result of exploreBundleResults) {
    const { bundleName, files, totalBytes } = result;

    // Check if source map contains only one file - this make result useless when exploring single bundle
    if (isSingleBundle) {
      const filenames = Object.keys(files).filter(
        filename => !SPECIAL_FILENAMES.includes(filename)
      );
      if (filenames.length === 1) {
        errors.push({
          bundleName,
          isWarning: true,
          code: 'OneSourceSourceMap',
          message: getErrorMessage({ code: 'OneSourceSourceMap', filename: filenames[0] }),
        });
      }
    }

    const unmappedBytes = files[UNMAPPED_KEY];
    if (unmappedBytes) {
      errors.push({
        bundleName,
        isWarning: true,
        code: 'UnmappedBytes',
        message: getErrorMessage({ code: 'UnmappedBytes', unmappedBytes, totalBytes }),
      });
    }
  }

  return errors;
}
