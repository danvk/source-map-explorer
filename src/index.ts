import { explore } from './api';
import { UNMAPPED_KEY, SOURCE_MAP_COMMENT_KEY, NO_SOURCE_KEY } from './explore';

export { explore, UNMAPPED_KEY, SOURCE_MAP_COMMENT_KEY, NO_SOURCE_KEY };

export default explore;

// Export all interfaces from index.ts to avoid type exports in compiled js code. See https://github.com/babel/babel/issues/8361

export type FileSizeMap = Record<string, number>;

export interface FileSizes {
  files: FileSizeMap;
  unmappedBytes: number;
  eolBytes: number;
  sourceMapCommentBytes: number;
  totalBytes: number;
}

export type ErrorCode =
  | 'Unknown'
  | 'NoBundles'
  | 'NoSourceMap'
  | 'OneSourceSourceMap'
  | 'UnmappedBytes'
  | 'InvalidMappingLine'
  | 'InvalidMappingColumn'
  | 'CannotSaveFile'
  | 'CannotCreateTempFile'
  | 'CannotOpenTempFile';

export type File = string | Buffer;

export type ReplaceMap = Record<string, string>;

export type OutputFormat = 'json' | 'tsv' | 'html';

/** Represents single bundle */
export interface Bundle {
  code: File;
  map?: File;
}

export interface ExploreOptions {
  /** Exclude "unmapped" bytes from the output */
  onlyMapped?: boolean;
  /** Exclude source map comment size from output */
  excludeSourceMapComment?: boolean;
  /** Output result as a string */
  output?: {
    format: OutputFormat;
    /** Filename to save output to */
    filename?: string;
  };
  /** Disable removing prefix shared by all sources */
  noRoot?: boolean;
  /** Replace "this" by "that" map */
  replaceMap?: ReplaceMap;
}

export interface ExploreResult {
  bundles: ExploreBundleResult[];
  /** Result as a string - either JSON, TSV or HTML */
  output?: string;
  errors: ExploreErrorResult[];
}

export interface ExploreBundleResult extends FileSizes {
  bundleName: string;
}

export interface ExploreErrorResult {
  bundleName: string;
  code: string;
  message: string;
  error?: NodeJS.ErrnoException;
  isWarning?: boolean;
}

export type BundlesAndFileTokens = (Bundle | string)[] | Bundle | string;

// TODO: Remove when https://github.com/mozilla/source-map/pull/374 is merged
declare module 'source-map' {
  export interface MappingItem {
    lastGeneratedColumn: number | null;
  }
}
