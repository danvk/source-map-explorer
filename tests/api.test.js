import { expect } from 'chai';
import fs from 'fs-extra';
import path from 'path';

import { explore, exploreBundlesAndWriteHtml, adjustSourcePaths } from '../src/api';

describe('Public API', function() {
  describe('explore', function() {
    it('should generate data when provided with js file with inline map', async function() {
      const actual = await explore('data/foo.min.inline-map.js');

      expect(actual).to.matchSnapshot();
    });

    it('should generate data when provided with file with referenced map', async function() {
      const actual = await explore('data/foo.min.js');

      expect(actual).to.matchSnapshot();
    });

    it('should generate data when provided with file with separated map file', async function() {
      const actual = await explore(
        'data/foo.min.no-map.js',
        'data/foo.min.no-map.separated.js.map'
      );

      expect(actual).to.matchSnapshot();
    });

    it('should generate data respecting onlyMapped and replace options', async function() {
      const actual = await explore('data/foo.min.js', 'data/foo.min.js.map', {
        onlyMapped: true,
        replace: { dist: 'hello' },
      });

      expect(actual).to.matchSnapshot();
    });

    it('should accept options passed as second or third argument', async function() {
      const actual3rdArg = await explore('data/foo.min.js', 'data/foo.min.js.map', {
        onlyMapped: true,
      });

      expect(actual3rdArg).to.matchSnapshot();

      const actual2ndArg = await explore('data/foo.min.js', { onlyMapped: true });

      expect(actual2ndArg).to.matchSnapshot();
    });

    it('should accept buffer with inline map', async function() {
      const actual = await explore(fs.readFileSync('data/foo.min.inline-map.js'));

      expect(actual).to.matchSnapshot();
    });

    it('should accept buffers with js and map', async function() {
      const actual = await explore(
        fs.readFileSync('data/foo.min.js'),
        fs.readFileSync('data/foo.min.js.map')
      );

      expect(actual).to.matchSnapshot();
    });

    it('should generate html', async function() {
      const actualBuffer = await explore(fs.readFileSync('data/foo.min.inline-map.js'), {
        html: true,
      });

      expect(actualBuffer)
        .to.have.property('html')
        .that.contains('<title>Buffer - Source Map Explorer</title>')
        .and.contains('"bar.js')
        .and.contains('"foo.js');

      const actualInline = await explore('data/foo.min.js', { html: true });

      expect(actualInline)
        .to.have.property('html')
        .that.contains('<title>data/foo.min.js - Source Map Explorer</title>')
        .and.contains('"bar.js')
        .and.contains('"foo.js');
    });

    it('should throw when specified file (js or map) not found', async function() {
      await expect(explore('data/something.js')).to.be.rejectedWith('no such file or directory');

      await expect(explore('data/foo.min.js', 'data/foo.min.js.maap')).to.be.rejectedWith(
        'no such file or directory'
      );
    });

    it('should trow when cannot locate sourcemap', async function() {
      await expect(explore('data/foo.min.no-map.js')).to.be.rejectedWith(
        'Unable to find a source map.'
      );
    });

    it('should throw when used with bad sourcemap', async function() {
      await expect(
        explore('data/foo.min.no-map.js', 'data/foo.min.no-map.bad-map.js.map')
      ).to.be.rejectedWith('Your source map only contains one source (foo.min.js)');
    });
  });

  describe('exploreBundlesAndWriteHtml', function() {
    function writeConfigToPath(writeConfig) {
      return writeConfig.path !== undefined
        ? `${writeConfig.path}/${writeConfig.fileName}`
        : writeConfig.fileName;
    }

    function expectBundleHtml(data) {
      expect(data).to.to.be.a('string');
      expect(data).to.have.string('<title>[combined] - Source Map Explorer</title>');
    }

    it('should explore multiple bundles and write a html file as specified in writeConfig', async function() {
      const writePath = path.resolve(__dirname, 'tmp');
      const writeConfig = {
        path: writePath,
        fileName: 'bundle-out.tmp.html',
      };

      await exploreBundlesAndWriteHtml(writeConfig, 'data/*.*');

      const data = fs.readFileSync(writeConfigToPath(writeConfig), 'utf8');

      expectBundleHtml(data);

      fs.removeSync(writePath);
    });

    it('should explore multiple bundles and write a html file to current directory if path is undefined in writeConfig', async function() {
      const writeConfig = { fileName: 'bundle-out.tmp.html' };

      await exploreBundlesAndWriteHtml(writeConfig, 'data/*.*');

      const data = fs.readFileSync(writeConfigToPath(writeConfig), 'utf8');

      expectBundleHtml(data);

      fs.removeSync(writeConfig.fileName);
    });
  });

  describe('adjustSourcePaths', function() {
    it('should factor out a common prefix', function() {
      [
        {
          sizes: { '/src/foo.js': 10, '/src/bar.js': 20 },
          expected: { 'foo.js': 10, 'bar.js': 20 },
        },
        {
          sizes: { '/src/foo.js': 10, '/src/foodle.js': 20 },
          expected: { 'foo.js': 10, 'foodle.js': 20 },
        },
      ].forEach(function({ sizes, expected }) {
        expect(adjustSourcePaths(sizes, true, [], [])).to.deep.equal(expected);
      });
    });

    it('should find/replace', function() {
      const expected = { '/dist/foo.js': 10, '/dist/foodle.js': 20 };

      const actual = adjustSourcePaths({ '/src/foo.js': 10, '/src/foodle.js': 20 }, false, {
        src: 'dist',
      });

      expect(actual).to.deep.equal(expected);
    });

    it('should find/replace with regexp', function() {
      const expected = { '/src/bar.js': 10, '/src/bar.le.js': 20 };

      const actual = adjustSourcePaths({ '/src/foo.js': 10, '/src/foodle.js': 20 }, false, {
        'foo.': 'bar.',
      });

      expect(actual).to.deep.equal(expected);
    });

    it('should find/replace with regexp, can be used to add root', function() {
      const expected = { '/bar/foo.js': 10, '/bar/foodle.js': 20 };

      const actual = adjustSourcePaths({ '/foo/foo.js': 10, '/foo/foodle.js': 20 }, false, {
        '^/foo': '/bar',
      });

      expect(actual).to.deep.equal(expected);
    });
  });
});
