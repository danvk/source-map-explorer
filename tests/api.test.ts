import { expect } from 'chai';
import fs from 'fs';
import rimraf from 'rimraf';
import snapshot from '@smpx/snap-shot-it';

import { explore, getBundles } from '../src/api';
import { setTestFolder, mockEOL } from './test-helpers';
import { BundlesAndFileTokens, ExploreOptions, Bundle } from '../src';

describe('api', () => {
  mockEOL();

  describe('explore', function() {
    setTestFolder();

    it('should generate data when provided with js file with inline map', async function() {
      const actual = await explore('data/inline-map.js');

      snapshot(actual);
    });

    it('should generate data when provided with file with referenced map', async function() {
      const actual = await explore('data/foo.min.js');

      snapshot(actual);
    });

    it('should generate data when provided with file with separated map file', async function() {
      const actual = await explore({
        code: 'data/no-map.js',
        map: 'data/no-map.js.map',
      });

      snapshot(actual);
    });

    it('should generate data respecting onlyMapped and replace options', async function() {
      const actual = await explore(
        { code: 'data/foo.min.js', map: 'data/foo.min.js.map' },
        {
          onlyMapped: true,
          replaceMap: { dist: 'hello' },
        }
      );

      snapshot(actual);
    });

    it('should accept buffer with inline map', async function() {
      const actual = await explore({ code: fs.readFileSync('data/inline-map.js') });

      snapshot(actual);
    });

    it('should accept buffers with js and map', async function() {
      const actual = await explore({
        code: fs.readFileSync('data/foo.min.js'),
        map: fs.readFileSync('data/foo.min.js.map'),
      });

      snapshot(actual);
    });

    describe('when output format specified', function() {
      it('should generate HTML from buffer', async function() {
        const result = await explore(
          { code: fs.readFileSync('data/inline-map.js') },
          { output: { format: 'html' } }
        );

        expect(result)
          .to.have.property('output')
          .that.contains('<title>Buffer - Source Map Explorer</title>')
          .and.contains('"bar.js')
          .and.contains('"foo.js');
      });

      it('should generate HTML from file', async function() {
        const result = await explore('data/foo.min.js', { output: { format: 'html' } });

        expect(result)
          .to.have.property('output')
          .that.contains('<title>data/foo.min.js - Source Map Explorer</title>')
          .and.contains('"bar.js')
          .and.contains('"foo.js');
      });

      it('should generate HTML from multiple files', async function() {
        const result = await explore(['data/foo.min.js', 'data/inline-map.js'], {
          output: { format: 'html' },
        });

        expect(result)
          .to.have.property('output')
          .that.contains('<title>[combined] - Source Map Explorer</title>')
          .and.contains('"bar.js')
          .and.contains('"foo.js');
      });

      it('should generate TSV', async function() {
        const result = await explore(['data/foo.min.js', 'data/inline-map.js'], {
          output: { format: 'tsv' },
        });

        snapshot(result.output);
      });

      it('should generate JSON', async function() {
        const result = await explore('data/foo.min.js', { output: { format: 'json' } });

        snapshot(result.output);
      });
    });

    describe(`when output filename specified`, function() {
      const tests: {
        name: string;
        filename: string;
        cleanPath?: string;
      }[] = [
        {
          name: 'should save html to file',
          filename: 'sme-result.html',
        },
        {
          name: 'should save html to file creating nested directories',
          filename: './tmp/nested/sme-result.html',
          cleanPath: './tmp',
        },
      ];

      tests.forEach(function({ name, filename, cleanPath }) {
        it(name, async function() {
          await explore('data/inline-map.js', {
            output: { format: 'html', filename },
          });

          expect(fs.existsSync(filename)).to.eq(true);

          const htmlContent = fs.readFileSync(filename).toString();
          expect(htmlContent).to.have.string(
            '<title>data/inline-map.js - Source Map Explorer</title>'
          );

          rimraf.sync(cleanPath || filename);
        });
      });
    });

    interface OnErrorTest {
      name: string;
      bundlesAndFileTokens: BundlesAndFileTokens;
      expectedErrorCode: string;
      options?: ExploreOptions;
    }

    describe('on error', () => {
      const bundleErrorTests: OnErrorTest[] = [
        {
          name: 'should throw when specified js file not found',
          bundlesAndFileTokens: 'data/something.js',
          expectedErrorCode: 'ENOENT',
        },
        {
          name: 'should throw when specified map file not found',
          bundlesAndFileTokens: { code: 'data/foo.min.js', map: 'data/foo.min.js.maap' },
          expectedErrorCode: 'ENOENT',
        },
        {
          name: 'should throw when cannot locate sourcemap',
          bundlesAndFileTokens: { code: 'data/no-map.js' },
          expectedErrorCode: 'NoSourceMap',
        },
      ];

      bundleErrorTests.forEach(function({
        name,
        bundlesAndFileTokens,
        options,
        expectedErrorCode,
      }) {
        it(name, async function() {
          try {
            await explore(bundlesAndFileTokens, options);
          } catch (result) {
            const error = result.errors[0];
            expect(error.code).to.equal(expectedErrorCode);
          }
        });
      });

      const appErrorTests: OnErrorTest[] = [
        {
          name: 'should throw when no bundles provided',
          bundlesAndFileTokens: [],
          expectedErrorCode: 'NoBundles',
        },
        {
          name: 'should throw when cannot save html to file',
          bundlesAndFileTokens: 'data/inline-map.js',
          options: { output: { format: 'html', filename: '?' } },
          expectedErrorCode: 'CannotSaveFile',
        },
      ];

      appErrorTests.forEach(function({ name, bundlesAndFileTokens, options, expectedErrorCode }) {
        it(name, async function() {
          try {
            await explore(bundlesAndFileTokens, options);
          } catch (error) {
            expect(error.code).to.equal(expectedErrorCode);
          }
        });
      });

      it('should not throw if at least one result is successful', async function() {
        const result = await explore(['data/foo.min.js', 'data/no-map.js']);

        expect(result.bundles.length).to.eq(1);
        expect(result.errors.length).to.eq(2);
      });

      it('should add error when used with bad sourcemap', async function() {
        const result = await explore('data/foo.min.no-map.bad-map.js');

        const error = result.errors[0];

        expect(error.code).to.equal('OneSourceSourceMap');
      });

      it('should add warning about unmapped bytes', async function() {
        const result = await explore('data/with-unmapped.js');

        const warning = result.errors[0];

        expect(warning.isWarning).to.equal(true);
        expect(warning.code).to.equal('UnmappedBytes');
      });
    });
  });

  interface GetBundlesTest {
    name: string;
    fileTokens: string[];
    expected: Bundle[];
  }

  describe('getBundles', function() {
    setTestFolder();

    const tests: GetBundlesTest[] = [
      {
        name: 'should expand glob',
        fileTokens: ['data/foo.min.js*'],
        expected: [
          {
            code: 'data/foo.min.js',
            map: 'data/foo.min.js.map',
          },
        ],
      },
      {
        name: 'should group code and map file into bundle',
        fileTokens: ['data/foo.min.js', 'data/foo.min.js.map'],
        expected: [
          {
            code: 'data/foo.min.js',
            map: 'data/foo.min.js.map',
          },
        ],
      },
      {
        name: 'should expand glob into all bundles in directory',
        fileTokens: ['data/*.*'],
        expected: [
          {
            code: 'data/foo.1234.js',
            map: 'data/foo.1234.js.map',
          },
          {
            code: 'data/foo.min.js',
            map: 'data/foo.min.js.map',
          },
          {
            code: 'data/foo.min.no-map.bad-map.js',
            map: 'data/foo.min.no-map.bad-map.js.map',
          },
          {
            code: 'data/inline-map.js',
            map: undefined,
          },
          {
            code: 'data/no-map.js',
            map: 'data/no-map.js.map',
          },
          {
            code: 'data/with-unmapped.js',
            map: 'data/with-unmapped.js.map',
          },
        ],
      },
      {
        name: 'should expand glob including .map files',
        fileTokens: ['data/foo.1*.js'],
        expected: [
          {
            code: 'data/foo.1234.js',
            map: 'data/foo.1234.js.map',
          },
        ],
      },
      {
        name: 'should expand glob into code and map files',
        fileTokens: ['data/foo.1*.js?(.map)'],
        expected: [
          {
            code: 'data/foo.1234.js',
            map: 'data/foo.1234.js.map',
          },
        ],
      },
      {
        name: 'should expand multiple globs',
        fileTokens: ['data/foo.1*.js', 'data/foo.mi?.js'],
        expected: [
          {
            code: 'data/foo.1234.js',
            map: 'data/foo.1234.js.map',
          },
          {
            code: 'data/foo.min.js',
            map: 'data/foo.min.js.map',
          },
        ],
      },
      {
        name: 'should support single file glob when inline map',
        fileTokens: ['data/inline*.js'],
        expected: [
          {
            code: 'data/inline-map.js',
            map: undefined,
          },
        ],
      },
    ];

    tests.forEach(function({ name, fileTokens, expected }) {
      it(name, function() {
        expect(getBundles(fileTokens)).to.deep.equal(expected);
      });
    });
  });
});
