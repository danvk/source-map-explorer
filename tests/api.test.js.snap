// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Public API explore should accept buffer with inline map 1`] = `
Object {
  "bundleName": "Buffer",
  "files": Object {
    "<unmapped>": 0,
    "dist/bar.js": 2854,
    "dist/foo.js": 137,
    "node_modules/browserify/node_modules/browser-pack/_prelude.js": 463,
  },
  "totalBytes": 3454,
  "unmappedBytes": 0,
}
`;

exports[`Public API explore should accept buffers with js and map 1`] = `
Object {
  "bundleName": "Buffer",
  "files": Object {
    "<unmapped>": 0,
    "dist/bar.js": 97,
    "dist/foo.js": 137,
    "node_modules/browserify/node_modules/browser-pack/_prelude.js": 463,
  },
  "totalBytes": 697,
  "unmappedBytes": 0,
}
`;

exports[`Public API explore should accept options passed as second or third argument 1`] = `
Object {
  "bundleName": "data/foo.min.js",
  "files": Object {
    "dist/bar.js": 97,
    "dist/foo.js": 137,
    "node_modules/browserify/node_modules/browser-pack/_prelude.js": 463,
  },
  "totalBytes": 697,
  "unmappedBytes": 0,
}
`;

exports[`Public API explore should accept options passed as second or third argument 2`] = `
Object {
  "bundleName": "data/foo.min.js",
  "files": Object {
    "dist/bar.js": 97,
    "dist/foo.js": 137,
    "node_modules/browserify/node_modules/browser-pack/_prelude.js": 463,
  },
  "totalBytes": 697,
  "unmappedBytes": 0,
}
`;

exports[`Public API explore should generate data respecting onlyMapped and replace options 1`] = `
Object {
  "bundleName": "data/foo.min.js",
  "files": Object {
    "hello/bar.js": 97,
    "hello/foo.js": 137,
    "node_modules/browserify/node_modules/browser-pack/_prelude.js": 463,
  },
  "totalBytes": 697,
  "unmappedBytes": 0,
}
`;

exports[`Public API explore should generate data when provided with file with referenced map 1`] = `
Object {
  "bundleName": "data/foo.min.js",
  "files": Object {
    "<unmapped>": 0,
    "dist/bar.js": 97,
    "dist/foo.js": 137,
    "node_modules/browserify/node_modules/browser-pack/_prelude.js": 463,
  },
  "totalBytes": 697,
  "unmappedBytes": 0,
}
`;

exports[`Public API explore should generate data when provided with file with separated map file 1`] = `
Object {
  "bundleName": "data/foo.min.no-map.js",
  "files": Object {
    "<unmapped>": 0,
    "dist/bar.js": 62,
    "dist/foo.js": 137,
    "node_modules/browserify/node_modules/browser-pack/_prelude.js": 463,
  },
  "totalBytes": 662,
  "unmappedBytes": 0,
}
`;

exports[`Public API explore should generate data when provided with js file with inline map 1`] = `
Object {
  "bundleName": "data/foo.min.inline-map.js",
  "files": Object {
    "<unmapped>": 0,
    "dist/bar.js": 2854,
    "dist/foo.js": 137,
    "node_modules/browserify/node_modules/browser-pack/_prelude.js": 463,
  },
  "totalBytes": 3454,
  "unmappedBytes": 0,
}
`;
